### ISSUE TITLE
SecondaryAxis.grid() fails silently

### ISSUE THREAD

**Issue by fohrloop** [matplotlib/matplotlib#18941]
This code is copy-pasted from [this SO question](https://stackoverflow.com/questions/64784834/draw-grid-line-on-secondaryaxis-matplotlib).

### Bug report

**Bug summary**

Creating a secondary axis and calling the `.grid()` method on it does not draw a grid. If it should not draw grid, it should throw an Exception. 

**Code for reproduction**

```python
import matplotlib.pyplot as plt
import numpy as np


def get_data_and_ticks():

    # generate dummy load duration curve
    dur = 2500
    y = np.random.normal(60, 30, dur + 1)
    y[::-1].sort()

    x = range(0, dur + 1)
    perticks = np.linspace(0, 1, 11)
    xticks = perticks * dur
    yticks = np.interp(xticks, range(0, dur + 1), y)

    return x, y, xticks, yticks


x, y, xticks, yticks = get_data_and_ticks()

# create figure object with axe object
fig, ax1 = plt.subplots(figsize=(10, 7))
ax1.plot(x, y)

# create second axis
color = "tab:blue"

ax2 = ax1.secondary_yaxis("right")
ax2.set_yticks(yticks)
ax2.set_yticklabels(labels=[f"{i*10}%" for i in range(11)])
ax2.tick_params(axis="y", color=color, labelcolor=color)

# draw grid lines on the secondaryaxis
# The problem is at this line
ax2.grid(color=color, which="major", linestyle="--")

plt.show()
```

**Actual outcome**

![image](https://user-images.githubusercontent.com/17479683/98970013-23808480-2518-11eb-8758-5e5c9748bd86.png)

**Expected outcome**

**Alternative A**
draw a figure like this:

![image](https://user-images.githubusercontent.com/17479683/98970115-3f842600-2518-11eb-8517-4634928705cf.png)

**Alternative B**
Raise an exception and tell the user that `SecondaryAxis.grid()` is not supported.

**Matplotlib version**

  * Operating system: Windows 10
  * Matplotlib version:  3.3.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): TkAgg
  * Python version: 3.8.6 64-bit

Same applies if a SecondaryAxis is created for `x` (`ax1.secondary_xaxis("top")`)

**Comment by jklymak** [comment]
`secondary_axis` is an axis, not an axes - the fact it is an axes subclass is an implementation detail.  If you want a full axes, and the ability to draw grid lines, then you would want to use `twin_axes`.  

I'm not particularly sure if there is an action item here.  We could whack-a-mole with all the possible `Axes` methods that should not be called with a secondary axis?

**Comment by anntzer** [comment]
Personally I think plotting (and other normal manipulations) on the secondary axes should be allowed(?) 
 (Implementation at https://github.com/matplotlib/matplotlib/pull/14463.)

**Comment by jklymak** [comment]
Well we could keep talking about that.  That is a twinned axes with extra linking between the axis-es.  But I'm not sure how we would separate and properly define all the subsequent interactions.

Take the above  - will the grid end up over or under a pcolormesh on the first axes?

**Comment by anntzer** [comment]
I don't see how this is a problem specific to secondary axes; after all twin axes basically have the same layering problem (everything on the twin will be drawn on top of everything on the base) and that doesn't seem to be a fatal flaw?
Also, if we really wanted to, we could have the base axes collect all children *both on itself and on the twin/secondary* and z-sort all of them in a single run (that's how mpl_toolkits's host/parasite axes are implemented).

**Comment by np-8** [comment]
### Implementation of a working SecondaryAxis.grid()?

Since it is quite easy to create the grid manually with 

```
xlim = ax1.get_xlim()
for y in ax2.get_yticks():
    ax1.plot(xlim, (y, y), ls="--", color=color, lw=0.5)
ax1.set_xlim(xlim)
```

wouldn't adding a `grid()` method on `SecondaryAxis` be doable? It already knows where the ticks are (depending on `self._orientation`, take `self.get_yticks()` or  `self.get_xticks()`), and the `xlim` can be read from the `self._parent.get_xlim()`. 

### Superclass of SecondaryAxis and the API

Could it be possible to split `_AxesBase` into two, where in another are the methods that are supported by `SecondaryAxis`? 

I personally find it a bit misleading if with a new library I create an object `obj = Object()`, and check with `dir(obj)` what methods there are available, and calling one of the methods does not seem to do what I expect, and does not give any warnings or exceptions either. 

### twinx vs secondary_yaxis

I understand there is the possibility to create "secondary axis" (or, more precisely, Axes) with twinx.  But, if the use case is like in this example
- Plot data to original axes `ax1` using original scales
- Create secondary axis/axes  `ax2` for having another, different set of xticks or yticks
- Add grid for the secondary axis/axes `ax2`

Then, isn't it a bit more compicated with twinx? I mean, you have to either scale your *data* to the new axis, too?  What do you think?

**Comment by jklymak** [comment]
@np-8  There are plusses and minuses to all those approaches.  And I agree that its inelegant to have a subclass not do anything useful for some methods.  If I were to think of alternatives I'd think of perhaps making SecondaryAxis a subclass of Axis, and attaching it to a hidden parent axes that the user doesn't really have access to.  

Splitting `_AxesBase` is possible but not terribly practical...

In your particular use case, I would simply make the blue axes the primary axes and the other one the secondary.

**Comment by np-8** [comment]
@jklymak Yeah I can see that there are always lots of pluses and minuses. By the way, even if `SecondaryAxis` would inherit from `Axis`, it could still be assumed that `SecondaryAxis.grid()` produces a grid (or throws Exception), since there is `Axis.grid()`. Of course, one possibility could be to  then create `_AxisBase` and inherit `Axis` and `SecondaryAxis` from it.

Thanks for the suggestion. I think that makes sense.

**Comment by anntzer** [comment]
https://github.com/matplotlib/matplotlib/issues/18941#issuecomment-726959792 still needs a decision.
